rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
Yeah, the code should definitely be moved into `handle_successful_job_execution`, regardless of whether we could simplify the code.

I re-read the code snippet I wrote and I still can't quite get why it wouldn't work. Dependent jobs should still be modifiable until `enqueue_dependents` is called and `pipeline.execute()` would ensure that `job` will be cleaned up in the same transaction.

I'll try to re-read this again this weekend. In the meantime, if you have the time, moving the code block into `handle_successful_job_execution` would be great :)

Thanks!